File: image-motion-studio/generate.py
#!/usr/bin/env python3
"""
Image Motion Studio — CLI
Usage:
    python generate.py <image_path> [output_dir]

Output:
    <name>_2.0s.mp4  — living cinematic clip with full optical & biological dynamics
"""

import os
import sys
import argparse
import subprocess
import shutil
from pathlib import Path

# Make sure backend modules are importable
CURRENT_DIR = Path(__file__).resolve().parent
BACKEND_DIR = CURRENT_DIR / "backend"
for p in (str(CURRENT_DIR), str(BACKEND_DIR)):
    if p not in sys.path:
        sys.path.insert(0, p)

from backend.depth import estimate_depth
from backend.renderer import run_pipeline

# ─────────────────────────────────────────────
# DEFAULT SETTINGS (Living Cinematic Engine)
# ─────────────────────────────────────────────
DEFAULTS = dict(
    duration            = 2.0,
    fps                 = 30,
    resolution          = "1080p",
    aspect_ratio        = "original",
    edge_fill           = "inpaint",

    # Camera Motion
    push_in             = 0.0,
    h_drift             = 0.0,
    v_drift             = 0.0,
    handheld            = 6.5,
    zoom_out            = 5.0,

    # Parallax / Depth
    depth_strength      = 15.0,
    foreground_separation = 10.0,

    # Bio-Motion & Physiological Dynamics
    breathing           = 10.0,
    watcher_sway        = 10.0,
    blink               = False,
    micro_saccades      = 2.5,
    edge_flutter        = 1.0,
    heartbeat_pulse     = 2.5,

    # Atmosphere & Optical Physics
    dust_particles      = 1.0,
    light_shift         = 2.0,
    film_grain          = 3.0,
    rack_focus          = 2.0,
    specular_shimmer    = 2.0,
    motion_blur         = 1.0,
    camera_shake        = 2.0,
)


def find_ffmpeg():
    """Return path to ffmpeg binary."""
    ffmpeg = shutil.which("ffmpeg")
    if not ffmpeg:
        raise RuntimeError("ffmpeg not found. Install with: brew install ffmpeg")
    return ffmpeg


def progress_bar(label: str):
    def cb(stage: str, pct: int):
        bar_len = 30
        filled = int(bar_len * pct / 100)
        bar = "█" * filled + "░" * (bar_len - filled)
        print(f"\r  {label} [{bar}] {pct:3d}%  {stage:<30}", end="", flush=True)
        if pct >= 100:
            print()
    return cb


def main():
    parser = argparse.ArgumentParser(
        description="Image Motion Studio CLI — generates living cinematic clips",
        formatter_class=argparse.RawDescriptionHelpFormatter,
        epilog="""
Examples:
  python generate.py photo.jpg
  python generate.py photo.jpg ./my_outputs
        """,
    )
    parser.add_argument("image", help="Path to input image (JPG/PNG/WEBP)")
    parser.add_argument("output_dir", nargs="?", default="outputs",
                        help="Output directory (default: outputs/)")
    args = parser.parse_args()

    image_path = Path(args.image).resolve()
    if not image_path.exists():
        print(f"❌ Image not found: {image_path}")
        sys.exit(1)

    output_dir = Path(args.output_dir).resolve()
    output_dir.mkdir(parents=True, exist_ok=True)

    temp_dir = output_dir / "temp"
    temp_dir.mkdir(parents=True, exist_ok=True)

    stem = image_path.stem

    print(f"\n🎬 Image Motion Studio — Living Cinematic Engine")
    print(f"   Image     : {image_path.name}")
    print(f"   Output dir: {output_dir}")
    print()

    # ── Step 1: Depth Estimation ──────────────────────────────────────────────
    print("1/2  Estimating depth (Depth Anything V2 Small, MPS)...")
    depth_path = str(temp_dir / f"{stem}_depth.png")
    depth_map = estimate_depth(str(image_path), depth_path)
    print(f"     ✅ Depth map: {depth_map.shape[1]}x{depth_map.shape[0]}")

    # ── Step 2: Render clip ─────────────────────────────────────────
    print(f"\n2/2  Rendering {DEFAULTS['duration']} sec video clip...")
    cb = progress_bar("     Render")

    output_file = run_pipeline(
        image_path        = str(image_path),
        depth_map         = depth_map,
        output_dir        = str(output_dir),
        temp_dir          = str(temp_dir),
        **DEFAULTS,
        progress_callback = cb,
    )

    final_path = output_dir / f"{stem}_{DEFAULTS['duration']}s.mp4"
    if Path(output_file).exists():
        shutil.move(output_file, final_path)

    print(f"     ✅ {final_path.name}  ({final_path.stat().st_size // 1024} KB)")

    # ── Done ─────────────────────────────────────────────────────────────────
    print(f"""
╔══════════════════════════════════╗
║  ✅  Done!                       ║
╠══════════════════════════════════╣
║  📹  {final_path.name}
║  🗂   {str(output_dir)}
╚══════════════════════════════════╝
""")

    # Clean up temp
    shutil.rmtree(str(temp_dir), ignore_errors=True)


if __name__ == "__main__":
    main()

File: image-motion-studio/generate.js
#!/usr/bin/env node
/**
 * Image Motion Studio — JavaScript CLI (generate.js)
 * Usage:
 *     node generate.js <image_path> [output_dir]
 * 
 * Output:
 *     <name>_2.0s.mp4 — living cinematic clip with full optical & biological dynamics
 */

'use strict';

const fs = require('fs');
const path = require('path');
const { estimateDepth } = require('./backend-node/depth');
const { runPipeline } = require('./backend-node/renderer');

// ─────────────────────────────────────────────
// DEFAULT SETTINGS (Living Cinematic Engine)
// ─────────────────────────────────────────────
const DEFAULTS = {
  duration: 2.0,
  fps: 30,
  resolution: '1080p',
  aspectRatio: 'original',
  edgeFill: 'inpaint',

  // Camera Motion
  pushIn: 0.0,
  hDrift: 0.0,
  vDrift: 0.0,
  handheld: 6.5,
  zoomOut: 5.0,
  cameraShake: 0.0,

  // Parallax / Depth
  depthStrength: 15.0,
  foregroundSeparation: 10.0,

  // Bio-Motion & Physiological Dynamics
  breathing: 10.0,
  watcherSway: 10.0,
  blink: false,
  microSaccades: 2.5,
  edgeFlutter: 1.0,
  heartbeatPulse: 2.5,

  // Atmosphere & Optical Physics
  dustParticles: 1.0,
  lightShift: 2.0,
  filmGrain: 3.0,
  rackFocus: 2.0,
  specularShimmer: 2.0,
  motionBlur: 1.0,
};

function progressBar(label) {
  return (stage, pct) => {
    const barLen = 30;
    const filled = Math.floor((barLen * pct) / 100);
    const bar = '█'.repeat(filled) + '░'.repeat(barLen - filled);
    process.stdout.write(`\r  ${label} [${bar}] ${String(pct).padStart(3)}%  ${stage.padEnd(30)}`);
    if (pct >= 100) {
      process.stdout.write('\n');
    }
  };
}

async function main() {
  const args = process.argv.slice(2);
  if (args.length === 0 || args[0] === '--help' || args[0] === '-h') {
    console.log(`
Image Motion Studio CLI (JavaScript) — generates living cinematic clips

Usage:
  node generate.js <image_path> [output_dir]

Examples:
  node generate.js photo.jpg
  node generate.js photo.jpg ./outputs
`);
    process.exit(0);
  }

  const imagePath = path.resolve(args[0]);
  if (!fs.existsSync(imagePath)) {
    console.error(`❌ Image not found: ${imagePath}`);
    process.exit(1);
  }

  const outputDir = path.resolve(args[1] || 'outputs');
  const tempDir = path.join(outputDir, 'temp');
  fs.mkdirSync(outputDir, { recursive: true });
  fs.mkdirSync(tempDir, { recursive: true });

  const stem = path.basename(imagePath, path.extname(imagePath));

  console.log(`\n🎬 Image Motion Studio — Living Cinematic Engine (100% JavaScript)`);
  console.log(`   Image     : ${path.basename(imagePath)}`);
  console.log(`   Output dir: ${outputDir}`);
  console.log();

  // ── Step 1: Depth Estimation ──────────────────────────────────────────────
  console.log('1/2  Estimating depth in pure JavaScript (Transformers.js ONNX)...');
  const depthPath = path.join(tempDir, `${stem}_depth.png`);
  const { depthMap, w, h, device } = await estimateDepth(imagePath, depthPath);
  console.log(`     ✅ Depth map: ${w}x${h} (${device})`);

  // ── Step 2: Render clip ─────────────────────────────────────────
  console.log(`\n2/2  Rendering ${DEFAULTS.duration} sec video clip...`);
  const cb = progressBar('     Render');

  const outputFile = await runPipeline({
    imagePath,
    depthMap,
    outputDir,
    tempDir,
    ...DEFAULTS,
    progressCallback: cb,
  });

  const finalPath = path.join(outputDir, `${stem}_${DEFAULTS.duration}s.mp4`);
  if (fs.existsSync(outputFile) && outputFile !== finalPath) {
    fs.renameSync(outputFile, finalPath);
  }

  const stats = fs.statSync(finalPath);
  console.log(`     ✅ ${path.basename(finalPath)} (${Math.round(stats.size / 1024)} KB)`);

  console.log(`
╔══════════════════════════════════╗
║  ✅  Done!                       ║
╠══════════════════════════════════╣
║  📹  ${path.basename(finalPath)}
║  🗂   ${outputDir}
╚══════════════════════════════════╝
`);

  // Clean up temp
  try {
    fs.rmSync(tempDir, { recursive: true, force: true });
  } catch {}
}

main().catch(err => {
  console.error('\n❌ Fatal error:', err);
  process.exit(1);
});
